torch/_decomp/decompositions.py

```diff
@@ -291,6 +382,59 @@ def huber_loss_backward(
     )
 
 
+@register_decomposition(aten.nll_loss_backward)
+def nll_loss_backward(grad_output: Tensor, self: Tensor, target: Tensor, weight: Optional[Tensor], reduction: int, ignore_index: int, total_weight: Tensor) -> Tensor:
+    assert 0 <= self.dim() <= 2, "input tensor should be 1D or 2D"
```

Comment:
we probably shouldn't actually use asserts for error checking, as they raise AssertionError which won't match the real deal (RuntimeError)

Reply:
Should we have some kind of “assert” operator? Might be difficult to model in a functional IR.

Reply:
We need an assert operator, not for this line (since we are expecting to always be rank specialized), but if we do anything with a dynamic shape we can't actually do the assert, we have to encode it into the IR. Functional IRs that don't care about validity can always drop it. Also cc @mruberry who has some thoughts on this.

Reply:
Having a special assertion operator, even if it only appears in traces, is an interesting idea. I don't know if we need one but it could be very helpful.

Reply:
I think having one is unavoidable. Asserts on rank may be evaluatable at trace/compile time, we *might* be able to get away with lifting dynamic shapes asserts up, but there other asserts like “assert x.min() > 0” in operators.

If we want to accurately model preconditions like that (and I’d argue we do!), then we need an assertion operator.

Reply:
We can understand the trace validity of expressions like `assert x.min() > 0` without modeling the assertion, though

Reply:
What does “understanding the trace validity mean”? Like, figuring out if you need to respecialize? (If so, I agree).

My point was that if we want to model the error checking a function has, we need to model the assert. It’s reasonable to say for now that “we’re just not gonna model the assert”, but eventually, we probably want to.

Reply:
> What does “understanding the trace validity mean”? Like, figuring out if you need to respecialize? (If so, I agree).

More generally, determining if you don't have a valid trace and so need to retrace, but yes.


> My point was that if we want to model the error checking a function has, we need to model the assert. It’s reasonable to say for now that “we’re just not gonna model the assert”, but eventually, we probably want to.

Idk. With the prims we have meta functions for all their error checking and shape logic. In the future we could even look at separating those two. We'll probably run those checks during trace acquisition.

Reply:
> We'll probably run those checks during trace acquisition.

How can we run a check like “x.min() > 0” during trace acquisition?

Reply:
Because we separate the error checking logic from the function's logic, so when we know the check is valid we just call the function without error checking, and if we don't know we run the check

Reply:
Sorry so in this case we have data-dependent control flow, so we'll almost certainly break the trace at data-dependent control flow. But if this were metadata-dependent control flow then my above statements apply (I think)

Reply:
> But if this were metadata-dependent control flow then my above statements apply (I think)

Yes I agree with this.

> so we'll almost certainly break the trace at data-dependent control flow.

What I'm saying is that we have an assert operator, we might be able to not break the trace at this kind of data-dependent control flow :) And I think we do have cases like that. At the very least, we might want to capture it into our trace simply to have the option to avoid CUDA syncs.

Reply:
As for the original discussion, I think I'm gonna leave the way it is now - any other solution is gonna introduce *actual* control flow, which I think is worse than raising an AssertionError instead of a RuntimeError.